src/main/java/org/template/Algorithm.java

```diff
@@ -850,29 +850,95 @@ public PredictedResult predict(NullModel model, Query query) {
       shouldFields.add(obj);
     }
-
-    String shouldScore =
-        "{\n" +
-            "   \"constant_score\": {\n" +
-            "  \"filter\": {\n" +
-            "  \"match_all\": {}\n" +
-            "},\n" +
-            "  \"boost\": 0\n" +
-            "  }\n" +
-            "}";
-    shouldFields.add(
-        new JsonParser().parse(shouldScore).getAsJsonObject()
-    );
-
-    return shouldFields;
-  }
-
-  private List<JsonElement> buildQueryMustNot(Query query, List<Event> boostable) {
-    return null;
-  }
-
-  private List<JsonElement> buildQuerySort() {
-    return null;
+      
+      String shouldScore =
+      "{\n" +
+      "   \"constant_score\": {\n" +
+      "  \"filter\": {\n" +
+      "  \"match_all\": {}\n" +
+      "},\n" +
+      "  \"boost\": 0\n" +
+      "  }\n" +
+      "}";
+      shouldFields.add(
+                       new JsonParser().parse(shouldScore).getAsJsonObject()
+                       );
+      
+      return shouldFields;
   }
 
+    /** Build sort query part */
+    private List<JsonElement> buildQuerySort(){
+        List<JsonElement> Seq = new ArrayList <JsonElement>();
+        if (recsModel == RecsModel.All || recsModel == RecsModel.BF){
+            Gson gson = new Gson();
+            List <JsonElement> sortByScore = new ArrayList <JsonElement>();
+            List <JsonElement> sortByRanks = new ArrayList <JsonElement>();
+            sortByScore.add(new JsonParser().parse("{\"_score\": {\"order\": \"desc\"}}"));
+            for(String fieldName:rankingFieldNames){
+                sortByRanks.add(new JsonParser().parse("{ \"$fieldName\": { \"unmapped_type\": \"double\", \"order\": \"desc\" } }"));
+            }
+            sortByScore.addAll(sortByRanks);
+            return sortByScore;
+        }
+        else{
+            return Seq;
+        }
+    }
+    
+    /** Build not must query part */
+    private List<JsonElement> buildQueryMustNot(Query query, List<Event> events){
+        List<JsonElement> mustNotFields = new ArrayList <JsonElement>();
+        Gson gson = new Gson();
+        JsonObject obj = new JsonObject();
+        JsonObject innerObj = new JsonObject();
+        
+        innerObj.addProperty("values", gson.toJson(getExcludedItems(events,query)));
+        obj.add("ids",innerObj);
+        obj.addProperty("boost", 0);
+        mustNotFields.add(obj);
+        
+        return mustNotFields;
+    }
+    
+    /** Create a list of item ids that the user has interacted with or are not to be included in recommendations */
+    private List<String> getExcludedItems(List<Event> events, Query query) {
+        List<Event> blacklistedItems = new ArrayList<Event>();
+        List<String> blacklistedStrings = new ArrayList<String>();
+        // either a list or an empty list of filtering events so honor them
+        for (Event event : events) {
+            if (blackListEvents.isEmpty()) {
+                if (event.equals(modelEventNames.get(0))) {
```

Comment:
When instantiating modelEventNames in Algorithm constructor, need to guarantee it's not null or empty.